### PR TITLE
Add document to setup PHAN for PHP applications

### PR DESCRIPTION
**Pull request by vishnu-kyatannawar**
**Summary**

- Setup `phan` with laravel basic app.
- Used basic `phan` configuration for laravel app.

# Below issues are fixed

**Issues**
- Currently we see a lot of Phan issues with basic laravel boilerplate code.
- Someone can take a deeper look at it and we can discuss further.

**Sample Issue**
```sh
app/Console/Kernel.php:8 PhanUndeclaredExtendedClass Class extends undeclared class \Illuminate\Foundation\Console\Kernel (Did you mean class \App\Console\Kernel or class \App\Http\Kernel)
app/Console/Kernel.php:16 PhanUndeclaredTypeParameter Parameter $schedule has undeclared type \Illuminate\Console\Scheduling\Schedule
app/Console/Kernel.php:16 PhanUnusedProtectedNoOverrideMethodParameter Parameter $schedule is never used
app/Console/Kernel.php:28 PhanUndeclaredMethod Call to undeclared method \App\Console\Kernel::load
app/Console/Kernel.php:30 PhanUndeclaredFunction Call to undeclared function \base_path()
app/Exceptions/Handler.php:8 PhanUndeclaredExtendedClass Class extends undeclared class \Illuminate\Foundation\Exceptions\Handler (Did you mean class \App\Exceptions\Handler)
app/Exceptions/Handler.php:35 PhanUndeclaredMethod Call to undeclared method \App\Exceptions\Handler::reportable
app/Exceptions/Handler.php:35 PhanUnusedClosureParameter Parameter $e is never used
app/Http/Controllers/Controller.php:10 PhanUndeclaredExtendedClass Class extends undeclared class \Illuminate\Routing\Controller (Did you mean class \App\Http\Controllers\Controller)
app/Http/Controllers/Controller.php:12 PhanUndeclaredTrait Class uses undeclared trait \Illuminate\Foundation\Auth\Access\AuthorizesRequests
app/Http/Controllers/Controller.php:13 PhanUndeclaredTrait Class uses undeclared trait \Illuminate\Foundation\Bus\DispatchesJobs
app/Http/Controllers/Controller.php:14 PhanUndeclaredTrait Class uses undeclared trait \Illuminate\Foundation\Validation\ValidatesRequests
app/Http/Kernel.php:7 PhanUndeclaredExtendedClass Class extends undeclared class \Illuminate\Foundation\Http\Kernel (Did you mean class \App\Console\Kernel or class \App\Http\Kernel)
app/Http/Kernel.php:19 PhanUndeclaredClassReference Reference to undeclared class \Fruitcake\Cors\HandleCors
app/Http/Kernel.php:21 PhanUndeclaredClassReference Reference to undeclared class \Illuminate\Foundation\Http\Middleware\ValidatePostSize
app/Http/Kernel.php:23 PhanUndeclaredClassReference Reference to undeclared class \Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull
app/Http/Kernel.php:33 PhanUndeclaredClassReference Reference to undeclared class \Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse
app/Http/Kernel.php:34 PhanUndeclaredClassReference Reference to undeclared class \Illuminate\Session\Middleware\StartSession
app/Http/Kernel.php:36 PhanUndeclaredClassReference Reference to undeclared class \Illuminate\View\Middleware\ShareErrorsFromSession
app/Http/Kernel.php:38 PhanUndeclaredClassReference Reference to undeclared class \Illuminate\Routing\Middleware\SubstituteBindings
app/Http/Kernel.php:44 PhanUndeclaredClassReference Reference to undeclared class \Illuminate\Routing\Middleware\SubstituteBindings
app/Http/Kernel.php:56 PhanUndeclaredClassReference Reference to undeclared class \Illuminate\Auth\Middleware\AuthenticateWithBasicAuth
```

**References**

Found this to be useful to resolve the above issues https://github.com/phan/phan/issues/504.

@sai-manohar  when you have time can you look into it?

We can either provide some stub files along with phan configuration to resolve the above issues if needed.

### CODE REVIEW

File: php-laravel-phan.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    'suppress_issue_types' => [
+		// The following two have been added to not highlight issues
+		// raised due to variables added for interface methods that are
+		// then not used by the interface implementation
+		'PhanUnusedPublicMethodParameter',
+		'PhanUnusedProtectedMethodParameter',
+		'PhanUnusedPrivateMethodParameter',
+		// Allow unused values in foreach
+		'PhanUnusedVariableValueOfForeachWithKey',
+		// FIXME: This should be fixed and removed, we should not swallow
```
Comment (Abijeet) [inline php-laravel-phan.md]: I think we can remove this. I believe it was specific for AIIO-X

**Comment by Abijeet** [comment]
There are some spelling mistakes in the PR description that I cannot change.

Regarding failures, its quite common to add stubs for classes, so feel free to do that.

File: php-laravel-phan.md (hunk 1 above)
Comment (vishnu-kyatannawar) [reply php-laravel-phan.md]: Done

File: php-laravel-phan.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
+# Introduction
```
Comment (Abijeet) [inline php-laravel-phan.md:1]: Do we also give recommendations regarding IDE etc?

**Comment by sai-manoharp** [comment]
Replying to https://github.com/OsmosysSoftware/lint-docs/pull/16#issue-1222776373

@vishnu-kyatannawar 

These issues are because vendor files which are required for the application to run are not included in `directory_list`. It should be fixed by https://github.com/OsmosysSoftware/lint-docs/commit/9b6f308dd8816691f3ffa0e7b88c88012dc7c761

I cloned [this](https://github.com/vishnu-kyatannawar/laravel-app), updated `phan` config and ran. Check screenshot for results. All the issues that are pointed are valid.

image

File: php-laravel-phan.md (hunk 2 above)
Comment (osm-vishnukyatannawar) [reply php-laravel-phan.md:1]: For IDE we thought to keep a different documentation.
For each technology which IDE to use, what extensions to install and workspace configuration settings to be used.
And any other related information in that.

File: php-laravel-phan.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+```
+
+# Usage
+
+Add script command to run `phan` in `composer.json` file.
+
+```json
+"scripts": {
+    // ..<existing scripts>
+    "phan": "phan",
```
Comment (osm-vishnukyatannawar) [inline php-laravel-phan.md]: Do you want to do something similar which you did with `phpcs`?

https://github.com/OsmosysSoftware/lint-docs/pull/21/files#diff-5ebe334af362550cb22dd6cc3db85d9e0c3eb108837c88538e9d8e601714bf2aR132

**Review by osm-vishnukyatannawar — APPROVED** [review]

File: php-laravel-phan.md (hunk 3 above)
Comment (sai-manoharp) [reply php-laravel-phan.md]: Updated this to use the project level configuration.

**Pull request merged**